daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/ElementKindParsers.scala

```diff
@@ -232,6 +234,24 @@ abstract class ChoiceDispatchCombinatorParserBase(rd: TermRuntimeData,
 
           if (pstate.processorStatus eq Success) {
             log(LogLevel.Debug, "Choice dispatch success: %s", parser)
+
+            // We usually rely on the sequence parser to set elements as final.
+            // But choices with scalar elements do not necessarily have a
+            // sequence surrounding them and so they aren't set final. In order
+            // to set these elements final, we do it here as well. We will
+            // attempt to walk the infoset after the PoU is discarded.
+            //
+            // Note that we must do a null check because it's possible there was
+            // a sequence, which figured out the element was final, walked it and
+            // it was removed.
+            val newLastChildNode = pstate.infoset.contents.lastOption
+            if (newLastChildNode != savedLastChildNode) {
+              val last = newLastChildNode.get
+              if (last != null) {
+                last.setFinal
```

Comment:
last.setFinal mutates the var _isFinal in trait DIFinalizable from false to true, so you probably should be calling it with parentheses like this:

last.setFinal(), not last.setFinal

If you search this pull request, you'll find three other places where setFinal is called without parentheses.